### PR TITLE
LBAC-33 :Add one or more access locations to a user

### PR DESCRIPTION
**Pull request by VANKINEENITAWRUN**
## Description of what I changed

- Made UI changes for the user to select a list of Locations.
- Locations are stored as userproperty with location UUID's concatenated.
- Made changes in Utils and in all `ServiceInterceptorAdvice` to remove those objects which do not belong to User's Location List. 

## Issue I worked on

see https://issues.openmrs.org/browse/LBAC-33

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

- [x] All new and existing **tests passed**.

- [x] My pull request is **based on the latest changes** of the master branch.

### CODE REVIEW

**Comment by HerbertYiga** [comment]
Did you  build the module and test these changes and found them working perfectly?

**Comment by VANKINEENITAWRUN** [comment]
@HerbertYiga  work on them need to make some changes in [this](https://github.com/openmrs/openmrs-module-referenceapplication/blob/master/omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java)  to allow the user to login  with multiple locations

**Comment by suthagar23** [comment]
Can you deploy the changes to https://modules-refapp.openmrs.org , and let me know to test it?

**Comment by dkayiwa** [comment]
@VANKINEENITAWRUN can you include the ticket id in your commit message as advised at? https://wiki.openmrs.org/display/docs/Pull+Request+Tips

**Comment by suthagar23** [comment]
One major concern here, In your ticket you have mentioned to work only for users. But this PR contains a lot of changes in other services (Patient/Encounter/..etc). 

Did you verify that all other features of this module can work without any issues regarding these changes?

File: api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java [hunk 1]
```diff
@@ -36,47 +38,52 @@ public static Location getPersonLocation(Person person) {
...
         }
         return false;
     }
 
-    public static Boolean compare(String value1, String value2) {
-        return (StringUtils.isNotBlank(value1) && StringUtils.isNotBlank(value2)) && value1.equals(value2);
-    }
-
     /**
      * Used to get the accessible location for the user. It will first get from the user property. if the user
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java]: If you change a method, then you need to think about the Java docs also

File: api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java [hunk 2]
```diff
@@ -36,47 +38,52 @@ public static Location getPersonLocation(Person person) {
         return null;
     }
 
-    public static Boolean doesPersonBelongToGivenLocation(Person person, PersonAttributeType personAttributeType, String sessionLocationUuid) {
+    public static Boolean doesPersonBelongToGivenLocations(Person person, PersonAttributeType personAttributeType, List<String> sessionLocationsLists) {
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java]: What is the use case of this change and why we need it? please clarify for me

Comment (VANKINEENITAWRUN) [reply api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java]: This method is used in person and patient serviceIntrepreterAdvice's to check whether they belong to current user Locations List.

**Comment by VANKINEENITAWRUN** [comment]
> One major concern here, In your ticket you have mentioned to work only for users. But this PR contains a lot of changes in other services (Patient/Encounter/..etc).

Yes, As the User is Having multiple Locations Access, we get a list of Locations in every InterceptorAdvice of the User so there should be filtered accordingly using the List.

**Comment by dkayiwa** [comment]
@suthagar23 any more comments for this?

File: api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java (hunk 1 above)
Comment (suthagar23) [reply api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java]: Have you seen this @VANKINEENITAWRUN ?

**Comment by suthagar23** [comment]
@dkayiwa Still some comments need to be addressed!

@VANKINEENITAWRUN Can you check the comments and follow JIRA guidelines also. 
Fix the conflicts also

File: api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java (hunk 1 above)
Comment (VANKINEENITAWRUN) [reply api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java]: yes fixed it

**Comment by VANKINEENITAWRUN** [comment]
> Fix the conflicts also

@suthagar23  fixed Merge Conflicts, can you please mention JIRA guidelines which I got missed so that I can catch up with them from next time.

**Comment by dkayiwa** [comment]
@suthagar23 what do you think of this?

**Pull request merged**